developer-docs/architecture/type-system.md

```diff
@@ -0,0 +1,260 @@
+often the case that we want to construct _composite_ values out of multiple `Input<T>`s.
+For example, consider `Input<Array<string>>`: a value of this type accepts either a
+`Array<string>` or an `Output<Array<string>>`, but does not accept a value of type
+`Array<Output<string>>`. In order to accept all three of these types, we need the type
+`Input<Array<Input<string>>>>`. The `inputShape` type constructor defines an algorithm
+for producing these sorts of values.
+
+```rust
+fn inputShape(T) {
+	match T {
+		_ => Input<T>,
+		Array<U> => Input<Array<inputShape(U)>>,
+		Map<U> => Input<Map<inputShape(U)>>,
+		Union<...U> => Union<map(...U, u => inputShape(u))>,
+		Promise<U> => Input<U>,
+		Output<U> => Input<U>,
+		Object<...P> => Input<Object<map(...P, (name, u) => (name, inputShape(u)))>>
+	}
+}
+```
+
+If we expand `Input<T>` into its underlying type, `Union<T, Output<T>>`, the types may be
+clearer:
+
+```rust
+fn inputShape(T) {
+	match T {
+		_ => Union<T, Output<T>>,
+		Array<U> => Union<Array<inputShape(U)>, Output<Array<inputShape(U)>>>,
+		Map<U> => Union<Map<inputShape(U)>, Output<Map<inputShape(U)>>>,
+		Union<...U> => Union<map(...U, u => inputShape(u))>,
+		Promise<U> => Union<U, Output<U>>,
+		Output<U> => Union<U, Output<U>>,
+		Object<...P> => Union<Object<map(...P, (name, u) => (name, inputShape(u)))>, Output<Object<map(...P, (name, u) => (name, inputShape(u)))>>>
+	}
+}
+```
+
+Resource input properties often use input-shaped types.
+
+## `outputShape(T)`
+
+Because the `Output<T>` metadata ([dependencies], [unknowns], and [secrets]) only applies
+to a single value, it is necessary to represent composite metadata using nested `Output<T>`
+types. Consider a variant of the `Array<string>` example from [`inputShape(T)`](#inputshapet):
+in order to produce an array where each element may be an output, we need to use the type
+`Output<Array<Output<string>>>`.
+
+The `outputShape` type constructor defines an algorithm for producing these sorts of values.
+
+```rust
+fn outputShape(T) {
+	match T {
+		_ => Output<T>,
+		Array<U> => Output<Array<outputShape(U)>>,
+		Map<U> => Output<Map<outputShape(U)>>,
+		Union<...U> => Union<map(...U, u => outputShape(u))>,
+		Promise<U> => Output<U>,
+		Output<U> => Output<U>,
+		Object<...P> => Output<Object<map(...P, (name, u) => (name, outputShape(u)))>>
+	}
+}
+```
+
+Resource output properties often use output-shaped types.
+
+Projecting output-shaped is a bit unwieldy, as values of these types often require a great
+deal of unwrapping as nesting depth increases.
+
+Instead of projecting these types in their fully-elaborated form, the various language SDKs
+tend to opt to project them as a simple `Output<T>` while using an internal representation
+for the concrete value that includes distinguished unknown values. This approach lets the
+SDKs to allow e.g. lifted property and element access into partially-known composite
```

Comment:
```suggestion
for the concrete value that includes distinguished unknown values. This approach allows the
SDKs to use e.g. lifted property and element access into partially-known composite
```

Reply:
I'm going to leave this one as-is.

Reply:
It doesn't read right as-is. Specifically the words: `This approach lets the SDKS to allow...` - Maybe just `This approach lets the SDKs allow...`

Reply:
ahhhh, yeah, the "to" in there is leftover from an earlier phrasing.